GH-715: Implemented the semantic highlighting support.

kittaakos opened the pull request:
This PR contains the following change:
 - Enables the semantic highlighting for the JDT LS.
 - Switches from LSP4J `0.4.0` to `0.5.0`.

Static:
![static_member](https://user-images.githubusercontent.com/1405703/43775743-0bb65de0-9a4e-11e8-851f-f983570a2ed9.gif)

Member vs. Parameter:
![member-vs-parameter](https://user-images.githubusercontent.com/1405703/43775754-15982082-9a4e-11e8-99a2-e44324e87857.gif)

Deprecated:
![deprecated](https://user-images.githubusercontent.com/1405703/43775764-1b2b1018-9a4e-11e8-9757-926b8937ed48.gif)

Auto (un)boxing:
![autoboxing](https://user-images.githubusercontent.com/1405703/43775783-2717f51c-9a4e-11e8-9ae0-18858f1bfe5f.gif)

Closes #715.

Signed-off-by: Akos Kitta <kittaakos@typefox.io>

fbricon commented:
@kittaakos thanks for the PR that looks promising! I'll review as soon as I can

kittaakos commented:
I will shrink the size of this PR once https://github.com/eclipse/lsp4j/pull/223 will be accepted and merged.

snjeza commented:
test this please

snjeza commented:
> I will shrink the size of this PR once eclipse/lsp4j#223 will be accepted and merged.

@kittaakos https://github.com/eclipse/lsp4j/pull/223 has been merged.

snjeza commented:
> The target platform is built from a fork of the original LSP4J source to enable the semantic highlighting capabilities.

@kittaakos Do I need to use any particular versions of VS Code?

kittaakos commented:
> Do I need to use any particular versions of VS Code?

Currently, VS Code does not expose a theming API. See [here](
https://github.com/Microsoft/vscode-languageserver-node/pull/367#issuecomment-401626625). You can try it out in [Eclipse Theia](https://github.com/theia-ide/theia); it [already uses](https://github.com/theia-ide/theia/blob/ca50055e84c2496701898fb0d879ef464a9d0d9e/packages/java/package.json#L63) the forked JDT language server.

fbricon commented:
Can we rename the org.eclipse.jdt.ls.core.internal.semantichighlight package to org.eclipse.jdt.ls.core.internal.highlighting? semantichighlight burns my eyes :-)

kittaakos commented:
> semantichighlight burns my eyes :-)

Sure. I do the updates here, once there is a new release of lsp4j. So that I can get rid of a bunch of classes, custom target platform files and so on. The release will happen either this or next week.

kittaakos commented:
I did the followings:
 - Switched to `0.5.0.M1` of LSP4J. (Adapted the `codeAction` and `documentSymbol` API to compile.)
 - Got rid of a bunch of obsolete utility classes and the custom target platform.
 - Renamed `org.eclipse.jdt.ls.core.internal.semantichighlight` to `org.eclipse.jdt.ls.core.internal.highlighting`.

fbricon commented:
@kittaakos could you extract the launcher/debugger stuff in a separate PR please, it's orthogonal to the issue and will be faster to get in

kittaakos commented:
> could you extract the launcher/debugger stuff in a separate PR please

@fbricon, I got rid of the non-highlighting related changes.

kittaakos commented:
Are there any updates, feedback on this PR?

svenefftinge commented:
Would be really nice to get this merged. It works nicely with Theia now. 
And we currently need to ship Theia with a fork.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/highlighting/SemanticHighlightingService.java:
```diff
@@ -0,0 +1,245 @@
+			infos.put(line, new SemanticHighlightingTokens.Token(character, length, scope));
+		}
+		//@formatter:off
+		return FluentIterable.from(infos.asMap().entrySet())
```
Java Streams instead?

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/highlighting/SemanticHighlightingService.java:
```diff
@@ -0,0 +1,245 @@
+import org.eclipse.lsp4j.VersionedTextDocumentIdentifier;
+import org.eclipse.lsp4j.util.SemanticHighlightingTokens;
+
+import com.google.common.base.Preconditions;
```
eclipse Assert?

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/highlighting/SemanticHighlightings.java:
```diff
@@ -0,0 +1,1021 @@
+/*******************************************************************************
```
if file was copied from jdt, then it should retain the original copyright from IBM.

Add Contributors section:
```xml
 *
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *     TypeFox - port to jdt.ls
```

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/highlighting/SemanticToken.java:
```diff
@@ -0,0 +1,117 @@
+/*******************************************************************************
+ * Copyright (c) 2018 TypeFox and others.
```
if file was copied from jdt, then it should retain the original copyright from IBM.

Add Contributors section:
```xml
 *
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *     TypeFox - port to jdt.ls
```

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/ClientPreferences.java:
```diff
@@ -172,4 +174,16 @@ public boolean isSupportsCompletionDocumentationMarkdown() {
+
+	public boolean isSemanticHighlightingSupported() {
+		TextDocumentClientCapabilities textDocument = capabilities.getTextDocument();
+		if (textDocument == null) {
```
use v3supported instead

kittaakos commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentLifeCycleHandler.java:
```diff
@@ -325,12 +341,33 @@ public void handleChanged(DidChangeTextDocumentParams params) {
-				IDocument document = JsonRpcHelpers.toDocument(unit.getBuffer());
-				edit.apply(document, TextEdit.NONE);
+
+				// Avoid any computation iff the `SemanticHighlightingService#isEnabled` is `false`.
```
Just for completeness with the [source](https://en.wikipedia.org/wiki/If_and_only_if):
"[...]_In logic and related fields such as mathematics and philosophy, if and only if (shortened **iff**) is a biconditional logical connective between statements._[...]"

Since it was not obvious whether it is a typo or not, I apply the recommended change. Thanks for the feedback.

kittaakos commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/highlighting/SemanticHighlighting.java:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	/**
+	 * Returns <code>true</code> iff the semantic highlighting consumes the semantic
```
That was the original code, but I do not mind changing the `iff` to `if`.

https://github.com/eclipse/eclipse.jdt.ui/blob/d41fa3326c5b75a6419c81fcecb37d7d7fb3ac43/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/SemanticHighlighting.java#L82

kittaakos commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java:
```diff
@@ -27,6 +27,7 @@
 import java.util.concurrent.Executors;
```
I do not understand this. Wha do you mean by this comment?

fbricon replied:
Just that those changes aren't necessary anymore, as they ve already been taken care of. Just needs to rebase against master

kittaakos commented:
Thank you for the review, @fbricon. I have updated this PR based on your feedback.

fbricon commented:
@jjohnstn is everything available in jdt.core now so that we can drop the copied files?

fbricon commented:
test this please

jjohnstn commented:
> @jjohnstn is everything available in jdt.core now so that we can drop the copied files?

Still in review.  https://git.eclipse.org/r/#/c/129807/
